text/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,511 @@
+`"C unwind"` function definition.
+
+Conversely, languages that support the native unwinding mechanism, such as C++,
+may be "sandwiched" between Rust frames, so that Rust `panic`s may safely
+unwind the C++ frames, if the Rust code declares both the C++ entrypoint and
+the Rust entrypoint using `"C unwind"`.
+
+## Other `unwind` ABI strings
+
+Because the `C` ABI is not appropriate for all use cases, we also introduce
+these `unwind` ABI strings, which will only differ from their non-`unwind`
+variants by permitting unwinding, with the same semantics as `"C unwind"`:
+
+* `"system unwind"` - available on all platforms
+* `"stdcall unwind"` and `"thiscall unwind"` - available only on platforms
+  where `"stdcall"` and `"thiscall"` are supported
+
+More `unwind` variants of existing ABI strings may be introduced, with the same
+semantics, without an additional RFC.
+
+## "Plain Old Frames"
+[POF-definition]: #plain-old-frames
+
+A "POF", or "Plain Old Frame", is defined as a frame that can be trivially
+deallocated: returning from or unwinding a POF cannot cause any
+observable effects. This means that POFs do not contain any pending destructors
+(live `Drop` objects) or `catch_unwind` calls.
+
+The terminology is intentionally akin to [C++'s "Plain Old Data"
+types][cpp-POD-definition], which are types that, among other requirements, are
+trivially destructible (their destructors do not cause any observable effects,
+and may be elided as an optimization).
+
+Rust frames that do contain pending destructors or `catch_unwind` calls are
+called non-POFs.
+
+Note that a non-POF may _become_ a POF, for instance if all `Drop` objects are
+moved out of scope, or if its only `catch_unwind` call is in a code path that
+will not be executed. The next section provides an example.
+
+[cpp-POD-definition]: https://en.cppreference.com/w/cpp/named_req/PODType
+
+## Forced unwinding
+[forced-unwinding]: #forced-unwinding
+
+This is a special kind of unwinding used to implement `longjmp` on Windows and
+`pthread_exit` in `glibc`. A brief explanation is provided in [this Inside Rust
+blog post][inside-rust-forced]. This RFC distinguishes forced unwinding from
+other types of foreign unwinding.
+
+Since language features and library functions implemented using forced
+unwinding on some platforms use other mechanisms on other platforms, Rust code
+cannot rely on forced unwinding to invoke destructors (calling `drop` on `Drop`
+types). In other words, a forced unwind operation on one platform will simply
+deallocate Rust frames without true unwinding on other platforms.
+
+This RFC specifies that, regardless of the platform or the ABI string (`"C"` or
+`"C unwind"`), any platform features that may rely on forced unwinding are
+undefined behavior if they cross non-[POFs][POF-definition]. For now, however,
+we do not specify the conditions required to use forced unwinding safely; we
+will specify this in [a future RFC][unresolved-questions].
```

Comment:
So this means only crossing POFs is necessary but not sufficient to make forced unwinding sound?

Reply:
Correct. This was the agreement we reached following the discussion about "cancelable" functions; currently, we expect to specify that functions are safely force-unwind-able if and only if they are POFs that are annotated `cancelable` (or whatever the annotation becomes). Of course, a non-POF that is annotated `cancelable` should cause a compiler error if possible.